Add option for including XFA and embedded files in metadata

ashkulz opened the pull request:
This patch series adds support for XFA and PDF attachments to be included in the PDF metadata, controlled via command-line options (disabled by default).

pwaller commented on src/main.cpp:
```diff
@@ -193,8 +197,29 @@ void dump_font_info(PDFDoc *doc) {
+
+  stream->reset();
+  while ((c = stream->getChar()) != EOF)
+    contents += (char)c;
```
Have you considered using [Stream::fillGooString](https://github.com/Distrotech/poppler/blob/cdb9cb2b64d70d73b2263c5aa65fafbf0ae4ec25/poppler/Stream.h#L134-L142) or an approach using `doGetChars` analogous to `fillGooString`? It may be more efficient than a per-byte function call.

ashkulz replied:
Done.

pwaller commented on src/main.cpp:
```diff
@@ -225,6 +250,56 @@ void dump_document_meta(const std::string filename, PDFDoc *doc,
+      }
+    }
+  } else {
+    packer.pack_nil();
```
Just an idea: rather than packing `nil` here, would it be cheap to pack some property of the form, for example `xfa.arrayGetLength()` if available else `0`?

ashkulz replied:
Done.

pwaller commented on README.md:
```diff
@@ -48,6 +48,8 @@ document (consecutive objects): <wire version : int> <metadata> <page>...
   "Pages": int,
   "FileName": str,
+  "XFA": dict,
+  "EmbeddedFiles": [<embedfile>...],
```
could you put `(or number of embedded files if --embedded-files not specified)`? (and likewise for `XFA`).

I didn't know about XFA before your PR, so it might be nice to have something describing what it is somewhere in the readme (particularly please use the word "form").

ashkulz replied:
Done in a separate section in README (and commit).

pwaller commented:
Hi @ashkulz, I'm back from holiday, so I'll be able to respond fairly quickly. Any progress on this?

ashkulz commented:
I'm on leave right now, will respond with updated PR by Friday.

pwaller commented:
Hi Again, just wanted to remind you about this.

ashkulz commented:
Sorry, I've been on medical leave for a month (viral fever). Will update by tomorrow.

pwaller commented:
No problem! Hope you are well recovered.

ashkulz commented:
@pwaller: addressed your review comments, changed the PR title, reformatted the code as requested and finally added a new section in the documentation.

Let me know if anything more needs to be done.

pwaller commented:
I think we are close to merging.

* Before we commit to this, can you have a last think about the wire format?
  
  Is it your perfect format?
  
  Once it goes in like this, it would be best to avoid changing. One thing that I might be concerned about - is it a good idea to have different types in the same structural position in the document? This sort of thing makes it hard to automatically map the structure onto a language type, e.g, a C++ structure/class. The field I'm referring to here is `EmbeddedFiles` being of type int or list depending on whether it is enabled.

* Would you mind breaking the code in `dump_document_meta` into `dump_..._xfa` and `dump_..._embedded_files`? (Feel free to pick reasonable names)

* Extremely minor nit: This project uses a stylistic convention of commit messages (and issue, PR titles) beginning with an upper case letter. Please update it if it's not too time consuming.

ashkulz commented:
I'd thought of it earlier, but then the options I evaluated were

1. always add a separate `EmbeddedFilesCount` key and set `EmbeddedFiles` key to nil if not enabled
1. always add a separate `EmbeddedFilesCount` key and add `EmbeddedFiles` key if not enabled
1. don't add the key `EmbeddedFiles` key if option not specified (makes the code a bit convoluted)
1. set `EmbeddedFiles` to nil if option not specified

I didn't like either of them, so chose the option I did. But you have a good point wrt data-type/auto-generated wrappers; can you let me know how well they deal with fields being optionally present? In a dynamic language/situation, you could introspect the value type. I'm inclined to choose option 4, same as what happens in case of XFA.

Will do the refactoring/message changes once the above gets confirmed.

pwaller commented:
(4) is reasonable. I'd be happy with that. You then at least avoid having a number type appear in the same location that an array type might appear which is the main issue.

I might have considered (1) or (slightly uglier, but functional) `{"EmbeddedFiles": {"N": 5, "Files": [...]}}`, where the latter can be absent or nil. I think you should go for 4 :)

ashkulz commented:
Done.

pwaller commented:
Thanks for your contribution.
